File: GP/Fase1/H.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define MOD 1000000007

int count_ways(set<string>& words, string& tattoo) {
    int n = tattoo.length();
    vector<int> dp(n+1, 0);
    dp[0] = 1;
    for (int i = 1; i <= n; i++) {
        for (int j = i-1; j >= 0; j--) {
            if (words.find(tattoo.substr(j, i-j)) != words.end()) {
                dp[i] += dp[j];
                dp[i] %= MOD;
            }
        }
    }
    return dp[n];
}

int main() {
    int n;
    cin >> n;
    set<string> words;
    for (int i = 0; i < n; i++) {
        string word;
        cin >> word;
        words.insert(word);
    }
    string tattoo;
    cin >> tattoo;
    cout << count_ways(words, tattoo) << endl;
}

File: GP/Fase1/H.py
MOD = 10**9 + 7

def count_ways(words, tattoo):
    n = len(tattoo)
    dp = [0] * (n + 1)
    dp[0] = 1
    for i in range(1, n+1):
        for j in range(i):
            if dp[j] and (tattoo[j:i] in words):
                dp[i] += dp[j]
                dp[i] %= MOD
    return dp[n]

n = int(input())
words = set()
for i in range(n):
    words.add(input().strip())
tattoo = input().strip()
print(count_ways(words, tattoo))
